int32 curr_pos1
int32 curr_pos2
int64 toff
float32 t